localstack/plugin/manager.py

```diff
@@ -0,0 +1,316 @@
+        _call_safe(
+            self.listener.on_load_before,
+            (plugin, load_args, load_kwargs),
+            "error while calling on_load_before",
+        )
+
+    def _fire_on_load_after(self, plugin, result):
+        _call_safe(
+            self.listener.on_load_after,
+            (plugin, result),
+            "error while calling on_load_after",
+        )
+
+    def _fire_on_load_exception(self, plugin, exception):
+        _call_safe(
+            self.listener.on_load_exception,
+            (plugin, exception),
+            "error while calling on_load_exception",
+        )
+
+
+class PluginContainer(Generic[P]):
+    """
+    Object to pass around the plugin state inside a PluginLoader.
+    """
+
+    name: str
+    lock: threading.Lock
+
+    plugin_spec: PluginSpec
+    plugin: P = None
+    load_value: Any = None
+
+    is_init: bool = False
+    is_loaded: bool = False
+
+    init_error: Exception = None
+    load_error: Exception = None
+
+
+class PluginManager(PluginLifecycleNotifierMixin, Generic[P]):
+    """
+    Manages Plugins within a namespace discovered by a PluginFinder. The default mechanism is to resolve plugins from
+    entry points using a StevedorePluginFinder.
+    """
+
+    namespace: str
+
+    load_args: Union[List, Tuple]
+    load_kwargs: Dict[str, Any]
+    listener: PluginLifecycleListener
+
+    def __init__(
+        self,
+        namespace: str,
+        load_args: Union[List, Tuple] = None,
+        load_kwargs: Dict = None,
+        listener: PluginLifecycleListener = None,
+        finder: PluginFinder = None,
+    ):
+        self.namespace = namespace
+
+        self.load_args = load_args or list()
+        self.load_kwargs = load_kwargs or dict()
+
+        self.listener = listener or PluginLifecycleListener()
+        self.finder = finder or StevedorePluginFinder(
+            self.namespace, self._fire_on_resolve_exception
+        )
+
+        self._plugin_index = None
+        self._init_mutex = threading.Lock()
+
+    def load(self, name: str) -> P:
+        container = self._require_plugin(name)
+
+        if not container.is_loaded:
+            self._load_plugin(container)
+
+        if container.init_error:
+            raise container.init_error
+
+        if container.load_error:
+            raise container.load_error
+
+        if not container.is_loaded:
+            # plugin should_load returned False
+            raise ValueError("plugin %s:%s is deactivated" % (self.namespace, name))
+
+        return container.plugin
+
+    def load_all(self, propagate_exceptions=False) -> List[P]:
+        plugins = list()
+
+        for name, container in self._plugins.items():
+            if container.is_loaded:
+                plugins.append(container.plugin)
+                continue
+
+            try:
+                plugin = self.load(name)
+                plugins.append(plugin)
+            except Exception:
+                if propagate_exceptions:
```

Comment:
Maybe the exception should be logged even if `propagate_exceptions` is `False`?

Reply:
good point. there may be duplicate logging in that case, since `load` re-raises previously logged exceptions. i added a debug line here and also a todo to maybe move the exception logging to the `load_all` method